desc/compute/bounce_integral.py

```diff
@@ -0,0 +1,1372 @@
+    shape = Z.shape
+    Z_ps = Z.reshape(Z.shape[0], Z.shape[1], -1)
+    f = [_interp1d_vec(Z_ps, knots, ff, method=method).reshape(shape) for ff in f]
+    B_sup_z = _interp1d_vec(Z_ps, knots, B_sup_z, method=method).reshape(shape)
+    # Specify derivative at knots for ≈ cubic hermite interpolation.
+    B = _interp1d_vec_with_df(Z_ps, knots, B, B_z_ra, method=method_B).reshape(shape)
+    pitch = pitch[..., jnp.newaxis, jnp.newaxis]
+    inner_product = jnp.dot(
+        _suppress_bad_nan(integrand(*f, B=B, pitch=pitch, Z=Z)) / B_sup_z,
+        w,
+    )
+    if check:
+        _assert_finite_and_hairy(Z, f, B_sup_z, B, B_z_ra, inner_product)
+    return inner_product
+
+
+_interpolatory_quadrature.__doc__ = _interpolatory_quadrature.__doc__.replace(
+    _delimiter, _repeated_docstring + _delimiter, 1
+)
+
+
+def _bounce_quadrature(
+    bp1,
+    bp2,
+    x,
+    w,
+    integrand,
+    f,
+    B_sup_z,
+    B,
+    B_z_ra,
+    pitch,
+    knots,
+    method="akima",
+    method_B="cubic",
+    check=False,
+):
+    """Bounce integrate ∫ f(ℓ) dℓ.
+
+    Parameters
+    ----------
+    bp1, bp2 : Array, Array
+        Each should have shape(P, S, bp1.shape[-1]).
+        The field line-following ζ coordinates of bounce points for a given pitch
+        along a field line. The pairs bp1[i, j, k] and bp2[i, j, k] form left
+        and right integration boundaries, respectively, for the bounce integrals.
+    x : Array, shape(w.size, )
+        Quadrature points in [-1, 1].
+
+    Returns
+    -------
+    result : Array, shape(P, S, bp1.shape[-1])
+        First axis enumerates pitch values. Second axis enumerates the field
+        lines. Last axis enumerates the bounce integrals.
+
+    """
+    errorif(x.ndim != 1 or x.shape != w.shape)
+    errorif(bp1.ndim != 3 or bp1.shape != bp2.shape)
+    pitch = jnp.atleast_2d(pitch)
+    S = B.shape[0]
+    if not isinstance(f, (list, tuple)):
+        f = [f]
+
+    def _group_grid_data_by_field_line(g):
+        msg = (
+            "Should have at most two dimensions, in which case the first axis "
+            "is interpreted as the batch axis, which enumerates the evaluation "
+            "of the function at particular pitch values."
+        )
+        errorif(g.ndim > 2, msg=msg)
+        return g.reshape(-1, S, knots.size)
+
+    f = map(_group_grid_data_by_field_line, f)
+    Z = affine_bijection_reverse(x, bp1[..., jnp.newaxis], bp2[..., jnp.newaxis])
+    # Integrate and complete the change of variable.
+    result = _interpolatory_quadrature(
+        Z, w, integrand, f, B_sup_z, B, B_z_ra, pitch, knots, method, method_B, check
+    ) * grad_affine_bijection_reverse(bp1, bp2)
+    assert result.shape == (pitch.shape[0], S, bp1.shape[-1])
+    return result
+
+
+_bounce_quadrature.__doc__ = _bounce_quadrature.__doc__.replace(
+    _delimiter, _repeated_docstring + _delimiter, 1
+)
+
+
+def bounce_integral(
+    eq,
```

Comment:
im coming around on the idea of passing the equilibrium around to compute functions. At the end of the day it is just a pytree, so we could basically replace `params` in all the compute functions with the actual object. That could potentially simplify a lot of things, and would probably make it easier to do compute functions that require multiple objects, like coil-coil distance etc.

Reply:
I just changed this to no longer take in an eq object so that I could compute effective ripple stuff with existing compute api, but yea could be useful